### PR TITLE
[3.x] PopupMenu rework and enhancements

### PR DESCRIPTION
**Pull request by rsubtil**
Backport of #41640 for `3.x` (cc @EricEzaM)

From my tests I didn't encounter any major breaking changes but there might be details about the popup behavior that I've missed.

I've done my best to port the PR with minimal changes, but also avoid having to massively change the existing codebase to accommodate for it. The only major deviation I had to do was adding in the variable `max_size` for setting the size, since that variable comes from `Window`, which  `Popup` was refactored to be a child of it (#37317)

### CODE REVIEW

File: scene/gui/popup_menu.cpp [hunk 1]
```diff
@@ -315,64 +321,58 @@ void PopupMenu::_gui_input(const Ref<InputEvent> &p_event) {
...
+				if (items[over].submenu != "") {
+					_activate_submenu(over);
+					return;
 				}
+				activate_item(over);
 			}
 		}
 
-		//update();
+		//control->update();
```
Comment (EricEzaM) [inline scene/gui/popup_menu.cpp:375]: Commented code - remove if not needed.

**Review by EricEzaM — COMMENTED** [review]
Did you take the changes straight from the diff of that PR? If so you may want to look at the current state of PopupMenu to see if anything else has been added to fix bugs etc.

**Comment by Ev1lbl0w** [comment]
It's a mix. I started by cherry-picking your commit and fixing conflicts, but then I compared with both `master` and `3.x`, and did a few changes, some to fix bugs and others to keep current `3.x` behavior.

File: doc/classes/PopupMenu.xml [hunk 2]
```diff
@@ -478,6 +478,9 @@
 		<member name="hide_on_state_item_selection" type="bool" setter="set_hide_on_state_item_selection" getter="is_hide_on_state_item_selection" default="false">
 			If [code]true[/code], hides the [PopupMenu] when a state item is selected.
 		</member>
+		<member name="max_size" type="float" setter="set_max_size" getter="get_max_size" default="0.0">
```
[the remark this answers is not in the document]
Comment (Ev1lbl0w) [reply doc/classes/PopupMenu.xml]: Fixed

File: scene/gui/popup_menu.cpp [hunk 3]
```diff
@@ -102,42 +100,65 @@ Size2 PopupMenu::get_minimum_size() const {
 		minsize.width += check_w;
 	}
 
+	if (max_size > 0) {
+		if (minsize.height > max_size) {
+			minsize.height = max_size;
+		}
```
[the remark this answers is not in the document]
Comment (Ev1lbl0w) [reply scene/gui/popup_menu.cpp]: Fixed

File: scene/gui/popup_menu.cpp (hunk 1 above)
Comment (Ev1lbl0w) [reply scene/gui/popup_menu.cpp]: Fixed

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**